bugwarrior/services/phab.py

```diff
@@ -178,7 +221,89 @@ def issues(self):
 
             extra = {
                 'project': project,
+                'projects': [diff_metadata['projects'][p_phid] for p_phid in projects_from_repo],
                 'type': 'pull_request',
                 #'annotations': self.annotations(phid, issue)
             }
             yield self.get_issue_for_record(diff, extra)
+
+def fetch_issue_metadata(api, issues):
+    """Get projects related to a list of issues from Phab
+
+    :api: Phabricator api object
+    :issues: A list of phab issue tuples with form (issue_phid, issue)
+    :returns: A dictionary with form {project_phid: project_data}
+    """
+    proj_phid_lists = [i[1]['projectPHIDs'] for i in issues]
+    return fetch_all_related_projects(api, proj_phid_lists)
+
+def fetch_diff_metadata(api, diffs):
+    """Get repo and projects related to a list of issues from Phab
+
+    :api: Phabricator api object
+    :diffs: A list of phab diffs
+    :returns: A dictionary with form {
+        'repo_binds': {repo_phid: [project_phid, ...]},
+        'projects': {project_phid: project_data}
+    }
+    """
+    repo_phids = [d['repositoryPHID'] for d in diffs]
+    repo_phids = [p for p in set(repo_phids)] # make unique
+
+    # We could try to also use the diff['auxiliary']['phabricator:projects']
+    # here, which the diff relevancy filtering code tries to do,
+    # but I haven't seen this information actually be set in any real
+    # calls to the phab conduit API
+
+    repos = api.diffusion.repository.search(
+        constraints={'phids': repo_phids},
+        attachments={'projects': True}
+    )
+
+    repo_proj_phid = {}
+    for r in repos.data:
+        repo_phid = r['phid']
+        att = r.get('attachments', None)
+        if att is None:
+            repo_proj_phid[repo_phid] = None
+            continue
+        proj_att = att.get('projects', None)
+        if proj_att is None:
+            repo_proj_phid[repo_phid] = None
+            continue
+        repo_proj_phid[repo_phid] = proj_att.get('projectPHIDs', None)
+
+    projects = fetch_all_related_projects(api, repo_proj_phid.values())
+
+    return {
+        'repo_binds': repo_proj_phid,
+        'projects': projects
+    }
+
+def fetch_all_related_projects(api, proj_phid_lists):
+    phids = set()
+    for pl in proj_phid_lists:
+        if pl is None:
+            continue
+        for phid in pl:
+            phids.add(phid)
+
+    phids = [p for p in phids] # set to list
```

Comment:
No! I had just forgotten this was possible, it's been about 2 years since I last wrote any python.